### PR TITLE
xref physical types

### PR DESCRIPTION
**Pull request by nstarman**
Now that #11118 is in (🥳 ), it's time for followups. One of the more interesting followups came from discussion with @adrn on methods to annotate physical types for Quantity and the like. @namurphy's great #11204 is in and #11606 is done, so I thought I'd give this a go.

See https://astropy--11595.org.readthedocs.build/en/11595/ for the compiled docs.

Part of #11555 
Requires : ~#11618~ (done)

## What this does (edited)

- convert the name of each physical type in the table of physical types in the docs to an ReST link target. These can be referenced (with sphinx :ref:) both in the docs and in parameters of docstrings by...
- Adding the link target to the target map in ``docs/conf.py::numpydoc_xref_aliases``.
- An explanation is added to the glossary.rst
- implements a non-exhaustive addition of physical type annotations in the docs. Some may be wrong! 

### Example

Screen Shot 2021-04-27 at 17 46 04

Links to the appropriate row in 

Screen Shot 2021-04-27 at 17 46 31

And the best part is that the link just looks like ``... ['length']``!

## Subpackage review sign-off sheet

@adonath @adamginsburg @StuartLittlefair @astrojuanlu @hamogu @dhomeier @saimn @matteobachetti @eslavich @tboch @tomdonaldson @perrygreenfield @mwcraig ~@taldcroft~ ~@adrn~ ~@mhvk~ @pllim @bsipocz @eteq @larrybradley @Cadair @astrofrog @mcara @nden

Please check and add your github handle after an approving review
- [ ] convolution:
- [x] coordinates: @mhvk @adrn
- [ ] modeling:
- [ ] stats:
- [x] table: @taldcroft @mhvk
- [ ] timeseries:
- [x] uncertainties: @mhvk
- [x] units: @mhvk @adrn 
- [ ] utils:
- [ ] visualization:

### CODE REVIEW

File: astropy/units/physical.py
Comment (nstarman) [inline astropy/units/physical.py:440]:
```suggestion
del unit, ptype, label, csv_reader, f, unitstring
```

File: docs/conf.py
Comment (nstarman) [inline docs/conf.py:172]:
```suggestion
    "dimensionless": ":ref:`one <ptype-dimensionless >`",
```

File: astropy/units/physical_types.csv
Comment (nstarman) [inline astropy/units/physical_types.csv:2]:
```suggestion
.. _ptype-dimensionless: ,dimensionless,
```

**Comment by mhvk** [comment]
@nstarman - I haven't looked in detail yet, but about storage I wonder whether you considered repurposing the way units are represented in the docs, which is basically by automatically generating the docstring for each module (see `generate_units_summary` in `utils.py`). I think I'd prefer that over a `.csv` file. 

Regardless, maybe best to do that kind of docs change as a separate PR from using the physical types in the docs?

On that: would #10655 enable moving towards allowing `~astropy.units.Quantity['length']`?

**Comment by nstarman** [comment]
#not_v4.3

**Comment by mhvk** [comment]
See #11618 for just getting the physical types as a table in the docs.

**Comment by nstarman** [comment]
Great. I'll label this as requiring 11618, and wait until that is merged. This PR will expand upon 11618 to add ReST link targets in the table.

**Comment by nstarman** [comment]
Ok. Trying to figure out how to get rid of that ugly spare column in the physical types table, but the link targets are up and running!
Now for adding docstring references all over the place.

File: astropy/units/physical.py
Comment (mhvk) [inline astropy/units/physical.py:549]: Wouldn't it work to use a direct reference, as in `my custom text <target to text>`_? (or, in this context,
```
- `{name} <_ptype-{name.replace(' ', '_')}`_
```

Also, why replace space with underscore - spaces in links do seem to be allowed.

Comment (mhvk) [reply astropy/units/physical.py:549]: Sorry, I'm looking at the wrong part of the text; I meant use 
```
- _`{name} <ptype-{name.replace(' ', '_')}`
```
which may not in fact work as the only thing I did note is that one can have an in-test target using _`target`

Comment (nstarman) [reply astropy/units/physical.py:549]: I just tried this and a couple variations. I can't get any of them work. I don't think RST has inline link target functionality.

Comment (mhvk) [reply astropy/units/physical.py:549]: It does have it, but quite possibly it does not work in tables -
see https://docs.readthedocs.io/en/stable/guides/cross-referencing-with-sphinx.html
```
For example, an in-line target inside a paragraph:

We can also create _`in-line targets` within an element on your page,
allowing you to, for example, reference text *within* a paragraph.
```

Comment (nstarman) [reply astropy/units/physical.py:549]: I found https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#inline-internal-targets
which has a nice explanation how to do this and other link types.

The link target that works is  ``_`length` `` with reference ``numpydoc_xref_aliases['length'] = "length <length>"

What doesn't work is the MUCH cleaner option `` _`length <ptype-length>` ``  with reference ``numpydoc_xref_aliases['length'] = "length <ptype-length>". Dumping the raw physical type names, unprepended by a descriptive prefix is potentially restrictive elsewhere in the docs. I haven't checked, but I suspect this will cause problems if a section is ever titled "work" or "time" in ANY section of the astropy docs (or docs that use astropy in their intersphinx, depending on the setup).

The alternative that avoids this issue has the somewhat ugly hanging column (see example from opening comment). This is a current limitation of sphinx.

Comment (nstarman) [reply astropy/units/physical.py]: I've gone with the former `_`length` ` with reference ``numpydoc_xref_aliases['length'] = "length <'length'>" ``

Comment (nstarman) [reply astropy/units/physical.py]: GTG!

**Review by taldcroft — APPROVED** [review]
Table changes LGTM.

File: astropy/coordinates/builtin_frames/altaz.py
Comment (astrofrog) [inline astropy/coordinates/builtin_frames/altaz.py:56]: Can we put ['dimensionless'] or similar here?

Comment (nstarman) [reply astropy/coordinates/builtin_frames/altaz.py]: Done!

File: astropy/units/equivalencies.py
Comment (mhvk) [inline astropy/units/equivalencies.py:803]: If we're listing a unit, might as well be the regular one; I guess 'frequency' is OK too.

File: astropy/units/physical.py
Comment (mhvk) [inline astropy/units/physical.py:546]: What a wonderfully small change here!!!

**Review by mhvk — APPROVED** [review]
This looks great!  I found one addition I thought was not quite right - but may have missed some as I didn't look very carefully. In any case, even if there is another, it will still clarify the docstring.
Approving with that one change.

File: astropy/units/equivalencies.py
Comment (nstarman) [reply astropy/units/equivalencies.py:803]: Ah. thanks. I though I'd corrected all those. That's a mistake. It should be frequency. A future PR will enable units in xref.

**Comment by mhvk** [comment]
No reason not to get it in 4.3, I think.

File: astropy/units/physical.py
Comment (nstarman) [reply astropy/units/physical.py:546]: It was unexpected and quite welcome 😀.

File: astropy/units/equivalencies.py
Comment (nstarman) [reply astropy/units/equivalencies.py]: When that happens, km / s / Mpc would probably be the preferred annotation.

**Comment by namurphy** [comment]
This PR looks great!  I can't wait for us to be able to do the same thing for PlasmaPy.

File: docs/glossary.rst
Comment (adrn) [inline docs/glossary.rst:38]:
```suggestion
       For example, `` distance : quantity-like ['length'] ``
```

**Review by adrn — APPROVED** [review]
Very minor nitpick, otherwise ready to merge!

**Comment by adrn** [comment]
@namurphy your comment makes me wonder: @nstarman are the xref physical types here usable by affiliated packages through intersphinx?

**Comment by nstarman** [comment]
> @namurphy your comment makes me wonder: @nstarman are the xref physical types here usable by affiliated packages through intersphinx?

That's a good question. They should be according to https://docs.readthedocs.io/en/stable/guides/intersphinx.html. One of my goals is to improve the overall ease of intersphinxing with astropy and using the glossary and xref aliases. Part of that goal is to move some of this and other code to `sphinx_astropy`, and pre-set up the inter sphinx... but until then, something like the following might work for numpydoc_xref in affiliate packages:

1.  add astropy to intersphinx
2. map in your conf.py

```python
from astropy.units.physical import _units_and_physical_types
numpydoc_xref_physical_type_aliases = {}
for _, ptypes in _units_and_physical_types:
    ptypes = {ptypes} if isinstance(ptypes, str) else ptypes
    for ptype in ptypes:
        key = f"'{ptype}'"
        val = f":ref:`:ref: '{ptype}' <astropy:{ptype}>`"   # <== intersphinxed link here
        numpydoc_xref_physical_type_aliases[key] = val

numpydoc_xref_aliases.update(numpydoc_xref_physical_type_aliases)
```

> Would be good to remember that we should probably document this PR in two places: (1) in the documentation part of the contributing guide (here?) to remind documentation contributions to include physical types when quantities/units are involved, and (2) to guide affiliated package maintainers with how to use xref physical types in their own docs, either through intersphinx or by copy-pasting the relevant part of the conf.py file modified here. Make a new issue to keep track of these things (for a follow-up)?

👍    #11675

File: docs/glossary.rst
Comment (adrn) [inline docs/glossary.rst:38]: Hm, this didn't render properly... is it the spaces around the backticks? https://astropy--11595.org.readthedocs.build/en/11595/glossary.html#term-physical-type

```suggestion
       For example, ``distance : quantity-like ['length']``
```

Comment (adrn) [reply docs/glossary.rst:38]: Cool - feel free to push a fix with `[ci skip]`

Comment (nstarman) [reply docs/glossary.rst]: Yup. That was it.

Comment (nstarman) [reply docs/glossary.rst]: 🤦

**Comment by namurphy** [comment]
@nstarman — thank you!  I've got too much pandemic brain fog to think much about it at the moment, but that looks like a good way to go about it for the time being.  I'm hoping to look into this more after 4.3 is released.

A slight hesitation is that `_units_and_physical_types` is not part of Astropy's public API, so I'm wondering if it would be worthwhile to put a comment in `physical.py` that says not to change the name of that `dict` since it may end up being used in other packages for this purpose.

cc: @rocco8773

**Comment by nstarman** [comment]
> A slight hesitation is that _units_and_physical_types is not part of Astropy's public API, so I'm wondering if it would be worthwhile to put a comment in physical.py that says not to change the name of that dict since it may end up being used in other packages for this purpose.

@namurphy, my thought for other packages is to essentially move the short script (if @adrn or others confirm it works) in https://github.com/astropy/astropy/pull/11595#issuecomment-830204184 to `sphinx_astropy` and export ``numpydoc_xref_physical_type_aliases`` (perhaps renamed to include "astropy") into ``v1`` so that in 3rd party packages' ``conf.py`` it gets imported. Then anyone who wants the physical-type xrefs can do ``numpydoc_xref_aliases.update(numpydoc_xref_physical_type_aliases)``. If they wanted different names, e.g. `` ['astropy-length'] `` instead of `` ['length'] `` if they support more than one unit package, then that can be done by 
``numpydoc_xref_aliases.update({"'astropy-"+k[1:]: v for k, v in numpydoc_xref_physical_type_aliases.items()})``.
This would hide ``_units_and_physical_types`` and make xref plugin quite easy. I think implementing https://github.com/astropy/astropy/pull/11595#issuecomment-830204184 directly in 3rd party packages should remain an experimental implementation and the next release of `sphinx_astropy` should codify the implementation for 3rd party packages.
Alternatively, a good addition to ``physical.py`` might be to add each physical type to the module (like how cosmology realizations are added in ``cosmology/realizations.py`` and have a set of all their names. I would 
``available_physical_types= _name_physical_mapping.keys()`` would conveniently stay updated as people define new physical types. Were this implemented, then that would change https://github.com/astropy/astropy/pull/11595#issuecomment-830204184 and this PR to use all publicly-scoped functions.

**Comment by adrn** [comment]
@namurphy I think using `_units_and_physical_types` is a temporary solution: we would ideally make it easier for other packages to get access to the physical type definitions directly in the future. So maybe better to put the comment in *your* docs :)

**Comment by adrn** [comment]
I think this is ready. Thanks again @nstarman - this is a great addition to the documentation!

**Pull request merged**
